### DISCUSSION TITLE
Update LXCs running Debian 11 to Debian 12

### DISCUSSION THREAD

**Discussion by tteck** [tteck/Proxmox#1498] [Announcements]
### Updating the LXC to Debian 12 manually _could_ potentially cause a loss of functionality.

Here's the approach I took to upgrade mine, but remember to have backups as a precautionary measure.

Copy & Paste in the LXC console

```
cat <<EOF >/etc/apt/sources.list
deb http://ftp.debian.org/debian bookworm main contrib
deb http://ftp.debian.org/debian bookworm-updates main contrib
deb http://security.debian.org/debian-security bookworm-security main contrib
EOF
apt-get update
DEBIAN_FRONTEND=noninteractive apt-get -o Dpkg::Options::="--force-confold" dist-upgrade -y
systemctl disable -q --now systemd-networkd-wait-online.service
```

```
rm -rf /usr/lib/python3.*/EXTERNALLY-MANAGED
apt-get autoremove --purge -y
```

**Comment by danbayliss** [comment]
Is this something we need to manually do for all containers or will the LXC updater script do it?

**Reply by tteck** [reply]
The update process will be performed manually since not everyone may prefer Debian 12 for various reasons. Once the Debian 12 template becomes available, you will be presented with the option to choose between Debian 11 or Debian 12 by accessing the Advanced settings in the scripts.

**Reply by danbayliss** [reply]
Thanks. I'll await the template being made available as without them i'd have not been able to get this far with Proxmox and I'll trust them more than I trust myself

**Reply by nicedevil007** [reply]
a suggestion on this. You could get the variable of the OS installed and if it is Debian 11, ask the user if he wants to upgrade to 12 during the LXC Updater script. Even better will be a question at the start of the script, if the user should be informed about the old debian installations or not, and if not the script will be run as we all know it right now?

**Comment by flix-one** [comment]
In case anyone is running into an issue with yarn install, the problem is the node.js version.
For example with NGINX Proxy Manager, Node.js version 16 is required.

You can easily use Node Version Manager --> [Install nvm](https://github.com/nvm-sh/nvm/blob/master/README.md#installing-and-updating)

Then use `nvm install 16.20.1` and `nvm use 16.20.1`
That solved the problem with NGINX Proxy Manager for me

**Reply by Nedlinin** [reply]
I ended up editing the script to insert NODE_OPTIONS=--openssl-legacy-provider in the package.json build script.  This allows using Node 18 as the build succeeds other than that issue.

**Comment by nicedevil007** [comment]
> ### Updating the LXC to Debian 12 manually _could_ potentially cause a loss of functionality.
> Here's the approach I took to upgrade mine, but remember to have backups as a precautionary measure.
> 
> Copy & Paste in the LXC console
> 
> ```
> cat <<EOF >/etc/apt/sources.list
> deb http://ftp.debian.org/debian bookworm main contrib
> deb http://ftp.debian.org/debian bookworm-updates main contrib
> deb http://security.debian.org/debian-security bookworm-security main contrib
> EOF
> apt-get update
> DEBIAN_FRONTEND=noninteractive apt-get -o Dpkg::Options::="--force-confold" dist-upgrade -y
> systemctl disable -q --now systemd-networkd-wait-online.service
> ```
> 
> ```
> rm -rf /usr/lib/python3.*/EXTERNALLY-MANAGED
> apt-get autoremove --purge -y
> ```

I would change the 5 lines to just one with

```python
sed -i 's/bullseye/bookworm/g' /etc/apt/sources.list
```

Thats the command that Proxmox is using in their wiki as well for just upgrading the bullseye stuff to bookworm in the sources.list.

**Reply by tteck** [reply]
The command `sed -i 's/bullseye/bookworm/g' /etc/apt/sources.list` replaces all occurrences of "bullseye" with "bookworm" in the `/etc/apt/sources.list` file, regardless of whether they are good or bad sources. On the other hand, using `cat <<EOF >/etc/apt/sources.list` ensures that only the sources considered good are included.

**Comment by JF62** [comment]
I got this error in several LXC Containers when I try to upgrade:

.
.
.
Processing triggers for ca-certificates (20230311) ...
Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
Errors were encountered while processing:
 linux-headers-6.1.0-10-amd64
 linux-headers-amd64
E: Sub-process /usr/bin/dpkg returned an error code (1)

does anyone have an idea what this can be?

**Reply by tteck** [reply]
try `apt --fix-broken install`

**Reply by JF62** [reply]
ok, where? In LXC Container? Host update with your script worked like a charm.
And... Before or after the error comes up? ( Backup installed already)

**Reply by tteck** [reply]
Which LXCs are experiencing problems?

**Reply by JF62** [reply]
unifi worked fine - pihole, Portainer, iobroker and influx/grafana all fail with this error. Wireguard I didn't try because I seen there are already not compatible with bookworm

**Reply by nicedevil007** [reply]
so because kernel 6.1 is not an official stable pve-kernel just reboot your host and select the default kernel there (bootmenu of proxmox) then try again and remove all old kernels like 6.1 with the scripts @tteck provided.

EDIT: what you can try as well is to unpin the 6.1 kernel with the efi boot tool provided by proxmox

https://pve.proxmox.com/wiki/Host_Bootloader

just to clear up, all LXCs are using the kernel from the host thats why you have to do the kernel stuff on the host first :)

**Reply by JF62** [reply]
I have already removed all old kernels and only Linux 6.2.16-4-pve at x86_64 is available and seen also in webmin as the used kernel in all LCX containers
EDIT:

root@pve:~# proxmox-boot-tool kernel list
Manually selected kernels:
None.

Automatically selected kernels:
6.2.16-4-pve

additional info: Could this be a reason for: https://forum.proxmox.com/threads/proxmox-ve-8-0-upgrade-error-while-building-linux-headers-6-1-0-10-amd64.130674/

**Comment by millie75** [comment]
I get the following when trying to update Adguard to Debian 12

alpine-adguard-sec login: root
Alpine-AdGuard LXC provided by https://tteck.github.io/Proxmox/

alpine-adguard-sec:~# cat <<EOF >/etc/apt/sources.list
> deb http://ftp.debian.org/debian bookworm main contrib
> deb http://ftp.debian.org/debian bookworm-updates main contrib
> deb http://security.debian.org/debian-security bookworm-security main contrib
> EOF
-ash: can't create /etc/apt/sources.list: nonexistent directory
alpine-adguard-sec:~# apt-get update
-ash: apt-get: not found
alpine-adguard-sec:~# DEBIAN_FRONTEND=noninteractive apt-get -o Dpkg::Options::="--force-confold" dist-upgrade -y
-ash: apt-get: not found
alpine-adguard-sec:~# systemctl disable -q --now systemd-networkd-wait-online.service^C

**Reply by tteck** [reply]
> Update LXCs running Debian 11 to Debian 12

This is to upgrade **Debian 11** to **Debian 12**, not **Alpine** to **Debian 12**

**Comment by liv-in-sky** [comment]
I tried with your  upgrade lines

the result in the nginx-lxc is:
![Image 033](https://github.com/tteck/Proxmox/assets/18462890/10139940-3b1f-4463-b771-d5f498ab9c3e)


but it seems that the nginx service is running **except** the web gui

error unser systemctl:

`
root@nginxproxymanager:~# systemctl status npm.service
x npm.service - Nginx Proxy Manager
     Loaded: loaded (/lib/systemd/system/npm.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Tue 2023-08-15 15:01:23 UTC; 7min ago
   Duration: 318ms
    Process: 344 ExecStartPre=mkdir -p /tmp/nginx/body /data/letsencrypt-acme-challenge (code=exite>
    Process: 345 ExecStart=/usr/bin/node index.js --abort_on_uncaught_exception --max_old_space_siz>
   Main PID: 345 (code=exited, status=1/FAILURE)
        CPU: 313ms

Aug 15 15:01:22 nginxproxymanager systemd[1]: npm.service: Main process exited, code=exited, status>
Aug 15 15:01:22 nginxproxymanager systemd[1]: npm.service: Failed with result 'exit-code'.
Aug 15 15:01:23 nginxproxymanager systemd[1]: npm.service: Scheduled restart job, restart counter i>
Aug 15 15:01:23 nginxproxymanager systemd[1]: Stopped npm.service - Nginx Proxy Manager.
Aug 15 15:01:23 nginxproxymanager systemd[1]: npm.service: Start request repeated too quickly.
Aug 15 15:01:23 nginxproxymanager systemd[1]: npm.service: Failed with result 'exit-code'.
Aug 15 15:01:23 nginxproxymanager systemd[1]: Failed to start npm.service - Nginx Proxy Manager.
`

**Reply by tteck** [reply]
Version 2.9.19 is already 9 months old (current version is 2.10.4). At this stage, I believe it's best to begin with a clean installation and subsequently stay updated with the latest versions.

**Reply by liv-in-sky** [reply]
what version do you mean ? the latest version of nginx seems to be: 1.24

**Reply by tteck** [reply]
[Nginx Proxy Manager](https://github.com/NginxProxyManager/nginx-proxy-manager/releases), not [Nginx](https://nginx.org/en/download.html)

**Reply by tteck** [reply]
The picture that you posted above shows the version

**Reply by liv-in-sky** [reply]
sorry - I realized it a couple of minutes before - do you have a link how to update ?

**Reply by tteck** [reply]
How did you originally install Nginx Proxy Manager?

**Reply by liv-in-sky** [reply]
with your proxmox  script

**Reply by tteck** [reply]
In that case, I recommend that you carefully go through the instructions. I still believe it's best to begin with a clean installation.

**Reply by liv-in-sky** [reply]
thanx a lot - I appriciate your work - I took an old backup - updatet to version 2.10.4 and used the lines fron the first post - now it seems to run  - external clients can connect und the webui is working

**Comment by andregarde** [comment]
I've been trying to update my Nginx Proxy Manager LXC on Debian 11 to 12 and keep running into problems.  I've been able to update to 2.10.4 on Debian 11 no problem.  I then execute the instructions in the OP to upgrade to Debian 12.  That seems to work okay, but when I visit the NPM homepage and try to login I get "Bad Gateway".  

Next, I tried re-running the upgrade script and it fails with this error:
```
✓ Stopped Services
 ✓ Cleaned Old Files
 ✓ Downloaded NPM v2.10.4
 ✓ Setup Enviroment
 - Building Frontend...
[ERROR] in line 137: exit code 1: while executing command yarn install --network-timeout=30000 &> /dev/null
```

I can restore a backup and thankfully that works for now.

**Reply by tteck** [reply]
Consider increasing your resources, including both Cores and Memory allocation.

**Reply by andregarde** [reply]
Increased the LXC to 4 GB RAM and 2 vCores, but still the same result.  In looking at the error I tried to execute the command that failed manually and this is what I get:

```
root@npm01:~/nginx-proxy-manager-2.10.4/frontend# yarn install --network-timeout=30000
yarn install v1.22.19
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning " > marionette.templatecache@1.0.0" has incorrect peer dependency "backbone@~1.3.3".
warning " > marionette.templatecache@1.0.0" has incorrect peer dependency "underscore@~1.8.3".
warning "marionette.templatecache > backbone.marionette@4.0.0-beta.1" has incorrect peer dependency "backbone@~1.3.3".
warning "marionette.templatecache > backbone.marionette@4.0.0-beta.1" has incorrect peer dependency "underscore@>= 1.8.3 <= 1.9.x".
warning "tabler-ui > bootstrap@4.5.1" has unmet peer dependency "popper.js@^1.16.1".
[4/4] Building fresh packages...
[-/5] ⠐ waiting...
[-/5] ⠐ waiting...
[-/5] ⠐ waiting...
[4/5] ⠐ node-sass
error /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass: Command failed.
Exit code: 1
Command: node scripts/build.js
Arguments: 
Directory: /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass
Output:
Building: /usr/bin/node /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/bin/node-gyp.js rebuild --verbose --libsass_ext= --libsass_cflags= --libsass_ldflags= --libsass_library=
gyp info it worked if it ends with ok
gyp verb cli [
gyp verb cli   '/usr/bin/node',
gyp verb cli   '/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/bin/node-gyp.js',
gyp verb cli   'rebuild',
gyp verb cli   '--verbose',
gyp verb cli   '--libsass_ext=',
gyp verb cli   '--libsass_cflags=',
gyp verb cli   '--libsass_ldflags=',
gyp verb cli   '--libsass_library='
gyp verb cli ]
gyp info using node-gyp@7.1.2
gyp info using node@18.13.0 | linux | x64
gyp verb command rebuild []
gyp verb command clean []
gyp verb clean removing "build" directory
gyp verb command configure []
gyp verb find Python Python is not set from command line or npm configuration
gyp verb find Python Python is not set from environment variable PYTHON
gyp verb find Python checking if "python3" can be used
gyp verb find Python - executing "python3" to get executable path
gyp verb find Python - executable path is "/usr/bin/python3"
gyp verb find Python - executing "/usr/bin/python3" to get version
gyp verb find Python - version is "3.11.2"
gyp info find Python using Python version 3.11.2 found at "/usr/bin/python3"
gyp verb get node dir no --target version specified, falling back to host node version: 18.13.0
gyp verb command install [ '18.13.0' ]
gyp verb install input version string "18.13.0"
gyp verb install installing version: 18.13.0
gyp verb install --ensure was passed, so won't reinstall if already installed
gyp verb install version is already installed, need to check "installVersion"
gyp verb got "installVersion" 9
gyp verb needs "installVersion" 9
gyp verb install version is good
gyp verb get node dir target node version installed: 18.13.0
gyp verb build dir attempting to create "build" dir: /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass/build
gyp verb build dir "build" dir needed to be created? /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass/build
gyp verb build/config.gypi creating config file
gyp verb build/config.gypi writing out config file: /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass/build/config.gypi
(node:1588) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
gyp verb config.gypi checking for gypi file: /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass/config.gypi
gyp verb common.gypi checking for gypi file: /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass/common.gypi
gyp verb gyp gyp format was not specified; forcing "make"
gyp info spawn /usr/bin/python3
gyp info spawn args [
gyp info spawn args   '/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/root/.cache/node-gyp/18.13.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/root/.cache/node-gyp/18.13.0',
gyp info spawn args   '-Dnode_gyp_dir=/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/root/.cache/node-gyp/18.13.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
Traceback (most recent call last):
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/gyp_main.py", line 51, in <module>
    sys.exit(gyp.script_main())
             ^^^^^^^^^^^^^^^^^
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 670, in script_main
    return main(sys.argv[1:])
           ^^^^^^^^^^^^^^^^^^
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 662, in main
    return gyp_main(args)
           ^^^^^^^^^^^^^^
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 629, in gyp_main
    [generator, flat_list, targets, data] = Load(
                                            ^^^^^
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 150, in Load
    result = gyp.input.Load(
             ^^^^^^^^^^^^^^^
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/pylib/gyp/input.py", line 3021, in Load
    LoadTargetBuildFile(
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/pylib/gyp/input.py", line 411, in LoadTargetBuildFile
    build_file_data = LoadOneBuildFile(
                      ^^^^^^^^^^^^^^^^^
  File "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/gyp/pylib/gyp/input.py", line 239, in LoadOneBuildFile
    build_file_contents = open(build_file_path, "rU").read()
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: invalid mode: 'rU' while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:291:12)
gyp ERR! System Linux 6.2.16-6-pve
gyp ERR! command "/usr/bin/node" "/root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
gyp ERR! cwd /root/nginx-proxy-manager-2.10.4/frontend/node_modules/node-sass
gyp ERR! node -v v18.13.0
```

Is it more feasible to create a new LXC and try to migrate the settings from one container to another?

**Reply by tteck** [reply]
> Updating the LXC to Debian 12 manually could potentially cause a loss of functionality.

Yes
